File: src/learning_assistant/problem_analyzer.py
from typing import Dict, List, Optional
import openai
from dataclasses import dataclass

@dataclass
class ProblemMetadata:
    title: str
    difficulty: str
    constraints: List[str]
    examples: List[Dict]
    input_format: str
    output_format: str
    
@dataclass
class AnalysisResult:
    patterns: List[str]
    approaches: List[Dict[str, str]]  # name -> description
    hints: List[str]
    learning_resources: List[str]
    complexity_requirements: Optional[Dict[str, str]]

class ProblemAnalyzer:
    def __init__(self, api_key: str):
        self.client = openai.Client(api_key=api_key)
        
    def extract_metadata(self, problem_text: str) -> ProblemMetadata:
        """Parse problem statement to extract structured metadata"""
        prompt = self._build_extraction_prompt(problem_text)
        response = self._get_completion(prompt)
        return self._parse_metadata(response)
        
    def analyze_patterns(self, metadata: ProblemMetadata) -> AnalysisResult:
        """Identify potential algorithmic patterns and approaches"""
        prompt = self._build_analysis_prompt(metadata)
        response = self._get_completion(prompt)
        return self._parse_analysis(response)

    def generate_hints(self, metadata: ProblemMetadata, user_progress: str) -> List[str]:
        """Generate contextual hints based on where user is stuck"""
        prompt = self._build_hint_prompt(metadata, user_progress)
        response = self._get_completion(prompt)
        return self._parse_hints(response)

    def _build_extraction_prompt(self, text: str) -> str:
        return f"""
        Analyze this coding problem and extract key information:
        
        {text}
        
        Extract and format:
        1. Constraints
        2. Input/Output formats
        3. Example cases
        4. Edge cases to consider
        """

    def _get_completion(self, prompt: str) -> str:
        response = self.client.chat.completions.create(
            model="gpt-4",
            messages=[{"role": "user", "content": prompt}]
        )
        return response.choices[0].message.content

    def _parse_metadata(self, response: str) -> ProblemMetadata:
        # Implementation of _parse_metadata method
        pass

    def _parse_analysis(self, response: str) -> AnalysisResult:
        # Implementation of _parse_analysis method
        pass

    def _parse_hints(self, response: str) -> List[str]:
        # Implementation of _parse_hints method
        pass

    def _build_analysis_prompt(self, metadata: ProblemMetadata) -> str:
        # Implementation of _build_analysis_prompt method
        pass

    def _build_hint_prompt(self, metadata: ProblemMetadata, user_progress: str) -> str:
        # Implementation of _build_hint_prompt method
        pass 

File: src/learning_assistant/problem_analyzer.js
import OpenAI from 'openai';

export class ProblemAnalyzer {
  constructor(apiKey) {
    this.client = new OpenAI({ 
      apiKey,
      dangerouslyAllowBrowser: true
    });
  }

  async extract_metadata(problemText) {
    const prompt = this._build_extraction_prompt(problemText);
    const response = await this._get_completion(prompt);
    return this._parse_metadata(response);
  }

  async analyze_patterns(metadata) {
    const prompt = this._build_analysis_prompt(metadata);
    const response = await this._get_completion(prompt);
    return this._parse_analysis(response);
  }

  async _get_completion(prompt) {
    const response = await this.client.chat.completions.create({
      model: "gpt-4",
      messages: [{ role: "user", content: prompt }]
    });
    return response.choices[0].message.content;
  }

  _build_extraction_prompt(text) {
    return `
      Analyze this coding problem and extract key information:
      
      ${text}
      
      Extract and format as JSON:
      {
        "title": "",
        "difficulty": "",
        "constraints": [],
        "examples": [],
        "input_format": "",
        "output_format": ""
      }
    `;
  }

  _parse_metadata(response) {
    return JSON.parse(response);
  }
} 